lib/sessions.js

```diff
@@ -0,0 +1,144 @@
+'use strict';
+
+const retrieveBSON = require('./connection/utils').retrieveBSON,
+  ReadPreference = require('./topologies/read_preference'),
+  BSON = retrieveBSON(),
+  Binary = BSON.Binary,
+  uuidV4 = require('./utils').uuidV4;
+
+/**
+ *
+ */
+class ClientSession {
+  constructor(topology, sessionPool, options) {
+    if (topology == null) {
+      throw new Error('ClientSession requires a topology');
+    }
+
+    if (sessionPool == null || !(sessionPool instanceof ServerSessionPool)) {
+      throw new Error('ClientSession requires a ServerSessionPool');
+    }
+
+    this.topology = topology;
+    this.sessionPool = sessionPool;
+    this.hasEnded = false;
+    this.serverSession = sessionPool.acquire();
+
+    options = options || {};
+    if (typeof options.initialClusterTime !== 'undefined') {
+      this.clusterTime = options.initialClusterTime;
+    }
+  }
+
+  /**
+   *
+   */
+  endSession(callback) {
+    if (this.hasEnded) {
+      return callback(null, null);
+    }
+
+    // TODO:
+    //   When connected to a sharded cluster the endSessions command
+    //   can be sent to any mongos. When connected to a replica set the
+    //   endSessions command MUST be sent to the primary if the primary
+    //   is available, otherwise it MUST be sent to any available secondary.
+    //   Is it enough to use: ReadPreference.primaryPreferred ?
+
+    this.topology.command(
+      'admin.$cmd',
+      { endSessions: 1, ids: [this.id] },
+      { readPreference: ReadPreference.primaryPreferred },
+      err => {
+        this.hasEnded = true;
+
+        // release the server session back to the pool
+        this.sessionPool.release(this.serverSession);
+
+        if (err) return callback(err, null);
+        callback(null, null);
+      }
+    );
+  }
+}
+
+Object.defineProperty(ClientSession.prototype, 'id', {
+  get: function() {
+    return this.serverSession.id;
+  }
+});
+
+/**
+ *
+ */
+class ServerSession {
+  constructor() {
+    this.id = { id: new Binary(uuidV4(), Binary.SUBTYPE_UUID) };
+    this.lastUse = Date.now();
+  }
+
+  /**
+   *
+   * @param {*} sessionTimeoutMinutes
+   */
+  hasTimedOut(sessionTimeoutMinutes) {
+    const idleTimeMinutes = Math.round(
+      (((Date.now() - this.lastUse) % 86400000) % 3600000) / 60000
```

Comment:
add a comment? not really clear what this formula represents.

Reply:
✅